### DISCUSSION TITLE
Usage with JSONAPI.org

### DISCUSSION THREAD

**Discussion by MartinMalinda** [CodeDredd/pinia-orm#643] [Q&A] [answered]
Hello!

Has anyone experimented with using this lib with a JSONAPI.org API? https://jsonapi.org/

I believe if anything, this explicit backend API should make the job easier for a client side ORM but the lib would need to provide some kind of adapter for it..

Ember data expects a JSONAPI.org backend by default and it worked pretty well there (https://api.emberjs.com/ember-data/release/classes/JSONAPIAdapter)

Thanks for any insights, tips!

**Comment by CodeDredd** [comment, marked as answer]
Hi,

can you make an example how you can imagine that could be used with pinia-orm ? It would help to discuess it if and how can be implemented. Thanks ❤️

**Reply by MartinMalinda** [reply]
@CodeDredd I'm double checking the docs and now I'm realizing that pinia-orm does not make requests to the backend right? So it's the responsibility of the consumer to make the fetch requests and then extract models from the json responses and do `save()`?

Based on my experience with ember data I assumed requests are being made. In ember data there's `fetch` and `find` and `peek` methods. Peek always looks only for client side data and is synchronous, find retreives local data if they exist otherwise does a request and fetch always does a request (these two return promises).

But if `pinia-orm` does not take responsibility for making fetch requests there's a possibility for creating a wrapper library, that would do that...

So there could be `pinia-orm-jsonapi` library that would also connect with the server for you 🤔 

Does this make sense?

**Reply by CodeDredd** [reply]
Yes that does make sense. I already started working on wrapper libaries for that case. I started with axios and then will continue. But you can use every lib you want for the api part. I am going also to update the docs what the best practice is to use that.
It would be to have a structure where you have repositories and models and always using the repository for the queries.
Files:
- repositories
  - BaseRepository
  - UserRepository (extending BaseRepository)
  - ... (extending BaseRepository)
- models
  - BaseModel
  - UserModel (extending BaseModel)
  - ... (extending BaseModel)

WIth this structure you will have the best to extend & maintain the models and repositories containing services.

So making a wrapper (i think a composable) for jsonapi would make sense, like making one for vue-query, or graphql, and so on.

**Reply by MartinMalinda** [reply]
Great! thank you! will watch out for a wrapper lib that I could use as a reference then 👀

**Reply by mahmoudawadeen** [reply]
@CodeDredd any updates regarding this?
